File: compiler/src/main/java/edu/cornell/cs/apl/viaduct/algebra/solver/ConstraintTerm.java
package edu.cornell.cs.apl.viaduct.algebra.solver;

import edu.cornell.cs.apl.viaduct.algebra.HeytingAlgebra;
import java.util.Map;

/** Terms that appear in constraints. */
public interface ConstraintTerm<A extends HeytingAlgebra<A>> {
  /** Returns the value of this term given an assignment of values to every variable in the term. */
  A getValue(Map<VariableTerm<A>, A> solution);

  /** Return the node that will represent this term in the constraint graph. */
  AtomicTerm<A> getNode();
}

File: compiler/src/main/kotlin/edu/cornell/cs/apl/viaduct/algebra/solver/ConstraintTerm.kt
package edu.cornell.cs.apl.viaduct.algebra.solver

import edu.cornell.cs.apl.viaduct.algebra.HeytingAlgebra

/** Terms that appear in constraints. */
interface ConstraintTerm<A : HeytingAlgebra<A>> {
    /** Returns the value of this term given an assignment of values to every variable in the term. */
    fun getValue(solution: ConstraintSolution<A>): A

    /** Return the node that will represent this term in the constraint graph. */
    val node: AtomicTerm<A>
}
